locale: also save X11 keyboard settings to /etc/vconsole.conf

yuwata opened the pull request:
Closes #24228.
Replaces #25412.

yuwata commented:
cc @n3rdopolis.

n3rdopolis commented:
Thanks! it seems to work cc @halfline

poettering commented on src/locale/localed-util.c:
```diff
@@ -56,13 +56,91 @@ static const char* systemd_language_fallback_map(void) {
+        xc->layout  = mfree(xc->layout );
+        xc->options = mfree(xc->options);
+        xc->model   = mfree(xc->model  );
+        xc->variant = mfree(xc->variant);
```
aligning the `=` makes sense to me, but aligning the `)` goes a little bit too far I think... dunno...

poettering commented on src/locale/localed-util.c:
```diff
@@ -56,13 +56,91 @@ static const char* systemd_language_fallback_map(void) {
+        return
+                isempty(xc->layout ) &&
+                isempty(xc->model  ) &&
+                isempty(xc->variant) &&
```
similar

poettering commented on src/locale/localed-util.c:
```diff
@@ -56,13 +56,91 @@ static const char* systemd_language_fallback_map(void) {
+        xc->layout  = (char*) empty_to_null(xc->layout);
+        xc->model   = (char*) empty_to_null(xc->model);
+        xc->variant = (char*) empty_to_null(xc->variant);
+        xc->options = (char*) empty_to_null(xc->options);
```
→ https://github.com/systemd/systemd/pull/25829

poettering commented on src/locale/localed-util.c:
```diff
@@ -421,30 +511,27 @@ int vconsole_convert_to_x11(Context *c) {
+                                                     .layout  = (char*) empty_or_dash_to_null(a[1]),
+                                                     .model   = (char*) empty_or_dash_to_null(a[2]),
+                                                     .variant = (char*) empty_or_dash_to_null(a[3]),
+                                                     .options = (char*) empty_or_dash_to_null(a[4]),
```
as above

poettering commented on src/locale/localed.c:
```diff
@@ -471,12 +417,37 @@ static int method_set_vc_keyboard(sd_bus_message *m, void *userdata, sd_bus_erro
+        if (convert) {
+                r = x11_read_data(c, m);
+                if (r < 0) {
+                        log_error_errno(r, "Failed to read x11 keyboard layout data: %m");
```
x11 is usually written X11

poettering commented on src/locale/localed.c:
```diff
@@ -471,12 +417,37 @@ static int method_set_vc_keyboard(sd_bus_message *m, void *userdata, sd_bus_erro
+                r = x11_read_data(c, m);
+                if (r < 0) {
+                        log_error_errno(r, "Failed to read x11 keyboard layout data: %m");
+                        return sd_bus_error_set(error, SD_BUS_ERROR_FAILED, "Failed to read x11 keyboard layout data");
```
ditto

poettering commented on src/locale/localed.c:
```diff
@@ -670,13 +663,13 @@ static int method_set_x11_keyboard(sd_bus_message *m, void *userdata, sd_bus_err
+                (void) vconsole_reload(sd_bus_message_get_bus(m));
 
         return sd_bus_reply_method_return(m, NULL);
 }
```
afaics this will make some file IO things fatal that previously weren't fatal, no?

I think it would make sense to make only issues with updating our internal idea what settings are in effect fatal. Any changes to the files on disk should probably be handled gracefully: logged about but otherwise ignored.

poettering replied:
rationale for the latter: we now write multiple files and we cannot possibly gurantee that either both updates worked or neither. unix fs apis simply have no concept that would allow us to implement that. hence I think we should update what we can, log about what we cannot, but return success as long as our in-memory state was updated.

poettering commented on src/locale/localed-util.c:
```diff
@@ -353,6 +373,25 @@ int vconsole_write_data(Context *c) {
+                if (r < 0)
+                        return r;
+
+                r = strv_env_assign(&l, "XKB_OPTIONS", empty_to_null(xc->options));
```
one of those days we might want to add strv_env_assign_many() or so with a varargs parm

poettering commented:
looks pretty good

yuwata commented:
@poettering Thank you for the review. All requests are addressed. PTAL.

poettering commented on src/locale/localed-util.c:
```diff
@@ -56,13 +56,91 @@ static const char* systemd_language_fallback_map(void) {
+                (!xc->layout  || string_is_safe(xc->layout) ) &&
+                (!xc->model   || string_is_safe(xc->model)  ) &&
+                (!xc->variant || string_is_safe(xc->variant)) &&
+                (!xc->options || string_is_safe(xc->options));
```
still don't like the `)` alignment

poettering commented on src/locale/localed-util.c:
```diff
@@ -56,13 +56,91 @@ static const char* systemd_language_fallback_map(void) {
+                streq_ptr(a->layout,  b->layout ) &&
+                streq_ptr(a->model,   b->model  ) &&
+                streq_ptr(a->variant, b->variant) &&
+                streq_ptr(a->options, b->options);
```
here neither

poettering commented on src/locale/localed-util.c:
```diff
@@ -310,17 +402,17 @@ int x11_write_data(Context *c) {
-        if (!isempty(c->x11_options))
-                fprintf(f, "        Option \"XkbOptions\" \"%s\"\n", c->x11_options);
+        if (!isempty(xc->options))
+                fprintf(f, "        Option \"XkbOptions\" \"%s\"\n", xc->options);
```
one of these days we should add proper escaping here

yuwata replied:
Escaping? We check the string with `string_is_safe()`, so I do not think escaping for it.

poettering commented on src/locale/localed.c:
```diff
@@ -471,12 +417,35 @@ static int method_set_vc_keyboard(sd_bus_message *m, void *userdata, sd_bus_erro
         r = vconsole_write_data(c);
         if (r < 0) {
                 log_error_errno(r, "Failed to set virtual console keymap: %m");
                 return sd_bus_error_set_errnof(error, r, "Failed to set virtual console keymap: %m");
```
should not be fatal i guess

n3rdopolis commented:
Seeing this merged would be pretty sweet, as I think it's still a blocker for Plymouth AFAIK

yuwata commented:
@poettering Addressed the comments about the spurious alignment. About 'escaping', see the above reply. I hope it is not a blocker of this PR. Upgrading the green label.

@n3rdopolis Sorry for late update.

keszybz commented on src/locale/localed-util.c:
```diff
@@ -351,44 +505,41 @@ static int read_next_mapping(const char* filename,
-                        !isempty(c->x11_model) ||
-                        !isempty(c->x11_variant) ||
-                        !isempty(c->x11_options);
+        xc = context_get_x11_context_safe(c);
```
I find the logic in this function very hard to follow. `context_get_x11_context_safe()` returns one of the contexts depending on their contents, and then we operate on it, even though the function is supposed to convert vconsole to x11. Dunno, maybe add an overview comment of what the logic is?

yuwata replied:
Sure.

keszybz reviewed (commented):
This is mostly questions…

yuwata commented on src/locale/localed-util.c:
```diff
@@ -86,6 +164,19 @@ void context_clear(Context *c) {
+
+X11Context *context_get_x11_context_safe(Context *c) {
+        return &ASSERT_PTR(c)->x11_from_xorg;
+}
```
Initially this is introduced as a mostly trivial function. But in a later commit, it will be extended.
Here, `_safe` does not mean the safety about the input, but it always return non-NULL value.

dtardon replied:
Maybe call it `context_ensure_x11_context`?

yuwata commented on src/locale/localed-util.c:
```diff
@@ -86,6 +164,19 @@ void context_clear(Context *c) {
+                return &c->x11_from_xorg;
+
+        return NULL;
+}
```
That's also for this function. See the comment below. The function will be extended in a later commit, and cannot make this 'const', as it will be used in `_safe()`.

yuwata commented on src/locale/localed-util.c:
```diff
@@ -167,14 +168,18 @@ void context_clear(Context *c) {
 static X11Context *context_get_x11_context(Context *c) {
         assert(c);
 
+        if (!x11_context_isempty(&c->x11_from_vc))
+                return &c->x11_from_vc;
```
Right, since this PR.

yuwata commented:
@keszybz Updated. Hopefully, all comments are addressed or answered. PTAL!

bluca commented:
@yuwata needs a rebase

bluca commented:
> @yuwata needs a rebase

done now

keszybz commented:
LGTM.

https://github.com/systemd/systemd/issues/24228#issuecomment-1207882785 has a nice overview of why this is useful.

poettering commented:
See → #26182, which renames the variables.

YHNdnzj commented on src/locale/localed-util.c:
```diff
@@ -136,6 +256,17 @@ int x11_read_data(Context *c, sd_bus_message *m) {
 
+        assert(c);
+
+        r = vconsole_read_data(c, m);
+        if (r < 0)
+                return r;
+
+        if (!x11_context_isempty(&c->x11_from_vc)) {
+                log_debug("XKB settings loaded from vconsole.conf, not reading xorg.conf.d/00-keyboard.conf.");
+                return 0;
+        }
```
@poettering I'm thinking if this is really desirable...

IMO we should respect more what is actually applied to X11, i.e `00-keyboard.conf` should have higher priority if that exists.

YHNdnzj replied:
And if the two get desynchronized, localed should be responsible for overwritting `vconsole.conf` based on `00-keyboard.conf`.
